core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+package zio.internal
+
+import zio.internal.ConcurrentWeakHashSet.Ref
+
+import java.lang.ref.WeakReference
+import scala.collection.mutable
+import java.util.concurrent.locks.ReentrantLock
+import java.lang.ref.ReferenceQueue
+import java.util.concurrent.atomic.AtomicInteger
+
+private[zio] object ConcurrentWeakHashSet {
+
+  private[internal] val MaxConcurrencyLevel: Int     = 1 << 16
+  private[internal] val MaxSegmentSize: Int          = 1 << 30
+  private[internal] val DefaultInitialCapacity: Int  = 16
+  private[internal] val DefaultLoadFactor: Float     = 0.75f
+  private[internal] val DefaultConcurrencyLevel: Int = 16
+
+  /**
+   * Reference node that links elements in the set with the same hash code
+   *
+   * @param hash
+   *   hash code of the element
+   * @param element
+   *   value passed to the weak reference
+   * @param queue
+   *   reference queue that will be used to track garbage collected elements
+   * @param nextRef
+   *   next reference node in the chain
+   * @tparam V
+   *   type of the element
+   */
+  protected class Ref[V](
+    val hash: Int,
+    element: V,
+    queue: ReferenceQueue[V],
+    var nextRef: Ref[V]
+  ) extends WeakReference[V](element, queue)
+
+  /**
+   * Access options for the update operation.
+   */
+  protected object UpdateOptions extends Enumeration {
```

Comment:
`Enumeration` is a dead Scala feature that folks avoid for good reason.

In its place, favor `sealed trait` extended by `case object` that appear in the companion object of `sealed trait`.

For example:

```scala
sealed trait UpdateOptions
object UpdateOptions {
  case object RestructureBefore extends UpdateOptions
  case object RestructureAfter extends UpdateOptions
  case object SkipIfEmpty extends UpdateOptions
  case object Resize extends Value
}
```

If you need to / from index support, you can add that manually, e.g.:

```scala
sealed trait UpdateOptions {
  def index: Int 
}
object UpdateOptions {
  case object RestructureBefore extends UpdateOptions {
    final val index = 0
  }
... 
  def fromIndex(index: Int): Option[UpdateOptions] = ...
}

```